File: mobile/lib/shared/models/achievement.dart
import 'package:reservatior/shared/enums/goal_type.dart';
import 'organization.dart';
import 'user.dart';

class Achievement {
  final String id;
  final String userId;
  final GoalType goalType;
  final int goalValue;
  final int currentValue;
  final bool isCompleted;
  final DateTime? completedAt;
  final int pointsReward;
  final String? bonusReward;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? organizationId;
  final Organization? organization;
  final User user;

  const Achievement({
    required this.id,
    required this.userId,
    required this.goalType,
    required this.goalValue,
    required this.currentValue,
    required this.isCompleted,
    this.completedAt,
    required this.pointsReward,
    this.bonusReward,
    required this.createdAt,
    required this.updatedAt,
    this.organizationId,
    this.organization,
    required this.user,
  });

  factory Achievement.fromJson(Map<String, dynamic> json) {
    return Achievement(
      id: json['id'] as String,
      userId: json['userId'] as String,
      goalType: GoalType.values.firstWhere((v) => v.name == json['goalType']),
      goalValue: json['goalValue'] as int,
      currentValue: json['currentValue'] as int,
      isCompleted: json['isCompleted'] as bool,
      completedAt: json['completedAt'] != null ? DateTime.parse(json['completedAt'] as String) : null,
      pointsReward: json['pointsReward'] as int,
      bonusReward: json['bonusReward'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
      organizationId: json['organizationId'] as String?,
      organization: json['organization'] != null ? Organization.fromJson(json['organization'] as Map<String, dynamic>) : null,
      user: User.fromJson(json['user'] as Map<String, dynamic>),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'goalType': goalType.name,
      'goalValue': goalValue,
      'currentValue': currentValue,
      'isCompleted': isCompleted,
      'completedAt': completedAt?.toIso8601String(),
      'pointsReward': pointsReward,
      'bonusReward': bonusReward,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'organizationId': organizationId,
      'organization': organization?.toJson(),
      'user': user.toJson(),
    };
  }

  Achievement copyWith({
    String? id,
    String? userId,
    GoalType? goalType,
    int? goalValue,
    int? currentValue,
    bool? isCompleted,
    DateTime? completedAt,
    int? pointsReward,
    String? bonusReward,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? organizationId,
    Organization? organization,
    User? user,
  }) {
    return Achievement(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      goalType: goalType ?? this.goalType,
      goalValue: goalValue ?? this.goalValue,
      currentValue: currentValue ?? this.currentValue,
      isCompleted: isCompleted ?? this.isCompleted,
      completedAt: completedAt ?? this.completedAt,
      pointsReward: pointsReward ?? this.pointsReward,
      bonusReward: bonusReward ?? this.bonusReward,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      organizationId: organizationId ?? this.organizationId,
      organization: organization ?? this.organization,
      user: user ?? this.user,
    );
  }
}

File: server/src/schemas/achievement.ts
import { z } from "zod";
import { GoalType } from "@reservatiorm/shared-types";

// Pagination schema
export const paginationSchema = z.object({
  page: z.string().regex(/^\d+$/).transform(Number).default("1"),
  limit: z.string().regex(/^\d+$/).transform(Number).default("20"),
  sortBy: z.string().default("createdAt"),
  sortOrder: z.enum(["asc", "desc"]).default("desc")
});

// Achievement filtering schema
export const achievementFilterSchema = z.object({
  userId: z.string().optional(),
  goalType: z.nativeEnum(GoalType).optional(),
  isCompleted: z.enum(["true", "false"]).optional().transform(val => val === "true"),
  organizationId: z.string().optional(),
  startDate: z.string().datetime().optional(),
  endDate: z.string().datetime().optional(),
  minPoints: z.string().regex(/^\d+$/).transform(Number).optional(),
  maxPoints: z.string().regex(/^\d+$/).transform(Number).optional()
});

// Achievement creation schema
export const achievementCreateSchema = z.object({
  userId: z.string().min(1, "User ID is required"),
  goalType: z.nativeEnum(GoalType),
  goalValue: z.number().min(0, "Goal value must be non-negative"),
  currentValue: z.number().min(0).optional(),
  isCompleted: z.boolean().optional(),
  completedAt: z.string().datetime().optional(),
  pointsReward: z.number().min(0).optional(),
  bonusReward: z.string().optional(),
  organizationId: z.string().optional(),
  description: z.string().optional(),
  badgeIcon: z.string().optional()
});

// Achievement update schema
export const achievementUpdateSchema = z.object({
  currentValue: z.number().min(0).optional(),
  isCompleted: z.boolean().optional(),
  completedAt: z.string().datetime().optional(),
  pointsReward: z.number().min(0).optional(),
  bonusReward: z.string().optional(),
  description: z.string().optional(),
  badgeIcon: z.string().optional()
});

// Progress update schema
export const progressUpdateSchema = z.object({
  currentValue: z.number().min(0, "Current value must be non-negative"),
  checkCompletion: z.boolean().default(true)
});

// Analytics query schema
export const analyticsQuerySchema = z.object({
  userId: z.string().optional(),
  organizationId: z.string().optional(),
  period: z.enum(["7d", "30d", "90d"]).default("30d"),
  groupBy: z.enum(["hour", "day"]).default("day")
});

// Leaderboard query schema
export const leaderboardQuerySchema = z.object({
  organizationId: z.string().optional(),
  period: z.enum(["7d", "30d", "90d", "all"]).default("all"),
  limit: z.string().regex(/^\d+$/).transform(Number).default("50"),
  goalType: z.string().optional()
});

// Export query schema
export const exportQuerySchema = z.object({
  userId: z.string().optional(),
  organizationId: z.string().optional(),
  goalType: z.nativeEnum(GoalType).optional(),
  isCompleted: z.enum(["true", "false"]).optional().transform(val => val === "true"),
  format: z.enum(["csv", "json"]).default("csv"),
  startDate: z.string().datetime().optional(),
  endDate: z.string().datetime().optional()
});

// Bulk creation schema
export const bulkCreateSchema = z.object({
  achievements: z.array(achievementCreateSchema).min(1, "At least one achievement is required")
});

// Combined schemas
export const getAchievementsSchema = paginationSchema.merge(achievementFilterSchema);
export const getUserAchievementsSchema = paginationSchema.pick({
  page: true, limit: true, sortBy: true, sortOrder: true
}).merge(z.object({
  goalType: z.nativeEnum(GoalType).optional(),
  isCompleted: z.enum(["true", "false"]).optional().transform(val => val === "true")
}));
